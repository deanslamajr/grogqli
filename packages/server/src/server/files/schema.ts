import shortid from 'shortid';
import editJsonFile from 'edit-json-file';
import { IntrospectionQuery } from 'graphql';
import {
  CreateTemporaryOperationRecording,
  SchemasMappingsInput,
} from '@grogqli/schema';
import fs from 'fs';

import {
  doesFileExist,
  getTemporarySchemaRecordingFilename,
  mapObjectToJsonFile,
  SCHEMA_FILE_VERSION,
  getSchemaRecordingFilePath,
  TEMP_SCHEMA_FILE_VERSION,
} from './';

type TemporarySchemaRecording = TemporarySchemaRecordingVersion1;
interface TemporarySchemaRecordingVersion1 {
  version: 1;
  hash: string;
  introspectionQuery: IntrospectionQuery;
}

type PersistTempSchemaRecording = (
  schemaIntrospectionResult: IntrospectionQuery
) => Promise<string>;

export const persistTempSchemaRecording: PersistTempSchemaRecording = async (
  schemaIntrospectionResult
) => {
  // TODO generate schema hash
  const schemaHash = shortid.generate();

  const newTempSchemaRecording: TemporarySchemaRecording = {
    version: TEMP_SCHEMA_FILE_VERSION,
    hash: schemaHash,
    introspectionQuery: schemaIntrospectionResult,
  };

  const schemaRecordingsPath = await getTemporarySchemaRecordingFilename(
    schemaHash
  );

  const newTempSchemaRecordingFile = editJsonFile(schemaRecordingsPath);

  if (!doesFileExist(newTempSchemaRecordingFile)) {
    mapObjectToJsonFile(newTempSchemaRecording, newTempSchemaRecordingFile);
    newTempSchemaRecordingFile.save();
  } else {
    console.log(`Schema with hash:${schemaHash} has already been recorded!`);
  }

  return schemaHash;
};

export type SchemaRecording = SchemaRecordingVersion1;
interface SchemaRecordingVersion1 {
  version: 1;
  id: string;
  name: string;
  hash: string;
  urls: string[];
  introspectionQuery: IntrospectionQuery;
}

export const getSchemaRecordingFile = async (
  schemaId: string
): Promise<SchemaRecording> => {
  const schemaRecordingFilePath = await getSchemaRecordingFilePath(schemaId);

  return JSON.parse(
    await fs.promises.readFile(schemaRecordingFilePath, 'utf8')
  );
};

type CreateNewSchemaRecording = (params: {
  schemaUrl: string;
  schemaHash: string;
}) => Promise<string>;
const createNewSchemaRecording: CreateNewSchemaRecording = async ({
  schemaHash,
  schemaUrl,
}) => {
  const tempSchemaRecordingsPath = await getTemporarySchemaRecordingFilename(
    schemaHash
  );

  let temporarySchemaRecordingFile: TemporarySchemaRecording;
  try {
    temporarySchemaRecordingFile = JSON.parse(
      await fs.promises.readFile(tempSchemaRecordingsPath, 'utf8')
    );
  } catch (error) {
    console.error(error);
    throw new Error(`
      Error while creating new schema recording from temporary schema recording:
      Cannot find a file for the given temporary schema recording hash:${schemaHash}
    `);
  }

  const generateUniqueSchemaIdAndCreateSchemaRecordingFile = async (): Promise<SchemaRecording> => {
    const newSchemaRecording: SchemaRecording = {
      version: SCHEMA_FILE_VERSION,
      id: shortid.generate(),
      name: 'tester schema', // TODO replace with user input
      hash: schemaHash,
      urls: [schemaUrl],
      introspectionQuery: temporarySchemaRecordingFile.introspectionQuery,
    };

    const schemaRecordingFilePath = await getSchemaRecordingFilePath(
      newSchemaRecording.id
    );

    const newSchemaRecordingFile = editJsonFile(schemaRecordingFilePath);

    if (doesFileExist(newSchemaRecordingFile)) {
      return generateUniqueSchemaIdAndCreateSchemaRecordingFile();
    } else {
      mapObjectToJsonFile(newSchemaRecording, newSchemaRecordingFile);
      newSchemaRecordingFile.save();
      return newSchemaRecording;
    }
  };

  const newSchemaRecording = await generateUniqueSchemaIdAndCreateSchemaRecordingFile();

  return newSchemaRecording.id;
};

export interface UpdatedSchemasMapping {
  schemaHash: string;
  schemaId: string;
}

type ConditionallyCreateOrUpdateSchemaRecordings = (
  schemaMappings: SchemasMappingsInput[]
) => Promise<UpdatedSchemasMapping[]>;
export const conditionallyCreateOrUpdateSchemaRecordings: ConditionallyCreateOrUpdateSchemaRecordings = async (
  schemaMappings
) => {
  return Promise.all(
    schemaMappings.map<Promise<UpdatedSchemasMapping>>(
      async ({
        opsRecordingsSchemaHash,
        targetSchemaId,
        opsRecordingsSchemaUrl,
      }) => {
        // handle NEW case
        if (targetSchemaId === 'NEW') {
          const newSchemaId = await createNewSchemaRecording({
            schemaUrl: opsRecordingsSchemaUrl,
            schemaHash: opsRecordingsSchemaHash,
          });
          return {
            schemaHash: opsRecordingsSchemaHash,
            schemaId: newSchemaId,
          };
        } else {
          // TODO
          // handle conditionally adding any new opsRecordingsSchemaUrl's to the mapped targetSchemaId
          return {
            schemaHash: opsRecordingsSchemaHash,
            schemaId: targetSchemaId,
          };
        }
      }
    )
  );
};
