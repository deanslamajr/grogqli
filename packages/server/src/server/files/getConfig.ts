import path from 'path';
import findUp from 'find-up';
import editJsonFile from 'edit-json-file';

const CONFIG_FILE_NAME = 'grogqli.json';

export interface GrogqliConfig {
  recordingsSaveDirectory: string;
  port: number;
  shouldDogFood: boolean;
}
const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: path.join(process.cwd(), 'grogqli'),
  port: 4000,
  shouldDogFood: false,
};

type GetGrogqliConfigValue = (configKey: keyof GrogqliConfig) => string;
let getGrogqliConfigValue: GetGrogqliConfigValue | undefined;

export const getConfig = async (): Promise<GetGrogqliConfigValue> => {
  // return early if the response is cached
  if (getGrogqliConfigValue) {
    return getGrogqliConfigValue;
  }

  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(
    async (directory) => {
      const hasConfig = await findUp.exists(
        path.join(directory, CONFIG_FILE_NAME)
      );
      return hasConfig ? directory : undefined;
    },
    { type: 'directory' }
  );

  let configFromFile: editJsonFile.JsonEditor;

  if (configLocation) {
    const configFileAbsolutePath = path.join(configLocation, CONFIG_FILE_NAME);
    configFromFile = editJsonFile(configFileAbsolutePath);
  }

  getGrogqliConfigValue = (configKey) => {
    if (configKey === undefined) {
      throw new Error(
        'Invalid use of getGrogqliConfigValue: must provide a key'
      );
    }
    let configValue: any;

    // First, try to resolve the value from the given config
    if (configLocation && configFromFile) {
      configValue = configFromFile.get(configKey);

      // Make 'recordingsSaveDirectory' an abs path
      if (
        configKey === 'recordingsSaveDirectory' &&
        configValue !== undefined
      ) {
        configValue = path.join(configLocation, configValue);
      }
    }

    // If config doesn't have a value, use the default
    if (configValue === undefined) {
      configValue = DEFAULT_CONFIG[configKey];
    }

    return configValue;
  };

  return getGrogqliConfigValue;
};
