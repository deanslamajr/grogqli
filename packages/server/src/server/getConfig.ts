import path from 'path';
import findUp from 'find-up';

const CONFIG_FILE_NAME = 'grogqli.json';

interface GrogqliConfig {
  recordingsSaveDirectory: string;
  recordingsFilename: string;
}

const DEFAULT_CONFIG: GrogqliConfig = {
  recordingsSaveDirectory: '/home/dhan/codes/grogqli',
  recordingsFilename: 'groql_rec'
};
export const getConfig = async (): Promise<GrogqliConfig> => {
  console.log('process.cwd()', process.cwd());
  // TODO
  // recursively try to open a file with filename === CONFIG_FILE_NAME
  // if file doesn't exist, try again in parent directory
  // if file isn't found at root directory, return DEFAULT_CONFIG
  const configLocation = await findUp(CONFIG_FILE_NAME);

  console.log('configLocation', configLocation)

  // TODO
  // if a config doesn't have a value defined for a particular lookup
  // use the value found in DEFAULT_CONFIG
  return DEFAULT_CONFIG;
}