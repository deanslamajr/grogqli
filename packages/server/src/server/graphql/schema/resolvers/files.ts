import editJsonFile from 'edit-json-file';
import fs from 'fs';

import { getConfig } from '../../../getConfig';

export const getRecordingFile = async (): Promise<editJsonFile.JsonEditor> => {
  const config = await getConfig();
  const recordingsSaveDirectory = config('recordingsSaveDirectory');
  if (!fs.existsSync(recordingsSaveDirectory)) {
    fs.mkdirSync(recordingsSaveDirectory);
  }

  return editJsonFile(
    `${recordingsSaveDirectory}/${config('recordingsFilename')}.json`
  );
};
