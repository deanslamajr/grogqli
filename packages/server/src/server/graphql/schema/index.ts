import { makeExecutableSchema } from '@graphql-tools/schema';
import { typeDefs } from '@grogqli/schema';

import { resolvers } from './resolvers';

const schema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

export default schema;
