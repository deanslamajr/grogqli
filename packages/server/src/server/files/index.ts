import editJsonFile from 'edit-json-file';
import fs from 'fs';
import path from 'path';
import { IntrospectionQuery } from 'graphql';

import { getConfig } from './getConfig';
import { HANDLER_SESSION_CREATED } from '../graphql/schema/resolvers/subscriptions/handlerSessionCreatedResolver';

const createDirIfDoesntExist = (dirPath: string) => {
  return fs.promises.mkdir(dirPath, { recursive: true });
};

export const doesFileExist = (file: editJsonFile.JsonEditor): boolean => {
  return Object.keys(file.read()).length > 0;
};

export const mapObjectToJsonFile = <T extends object>(
  object: T,
  file: editJsonFile.JsonEditor
): void => {
  Object.entries(object).forEach(([key, value]) => {
    file.set(key, value);
  });
};

// Directory structure
//
// /grogqli
//   /schemas
const SCHEMAS_FOLDER_NAME = 'schemas';
//     index.json
//     /<schemaId>
//       operations.json
export const OPERATIONS_NAME_TO_ID_MAPPING_FILENAME = 'operations.json';
export const OPERATIONS_NAME_TO_ID_MAPPING_VERSION = 1;
//       schema.json
const SCHEMA_FILENAME = 'schema.json';
//       types.json
const TYPES_NAME_TO_ID_MAPPING_FILENAME = 'types.json';
export const TYPES_NAME_TO_ID_MAPPING_VERSION = 1;
//   /operations
export const OPERATIONS_FOLDER_NAME = 'operations';
export const OPERATIONS_RECORDING_VERSION = 1;
//     someOperationId.json
//   /types
const TYPES_FOLDER_NAME = 'types';
//     <typeId>.json
export const TYPES_FILE_VERSION = 1;
//   /workflows
export const WORKFLOWS_FOLDER_NAME = 'workflows';
//     index.json
const WORKFLOWS_NAME_TO_ID_MAPPING_FILENAME = 'index.json';
export const WORKFLOWS_NAME_TO_ID_MAPPING_VERSION = 1;
//     <workflowId>.json
//   /local
const TEMP_FOLDER_NAME = 'local';
//     appState.json
//     /sessions
const HANDLER_SESSIONS_STATE_FOLDER_NAME = 'sessions';
//      index.json
const HANDLER_SESSION_STATE_FILENAME = 'index.json';
export const HANDLER_SESSIONS_STATE_FILE_VERSION = 1;
//      /schemas
const TEMP_SCHEMAS_FOLDER_NAME = 'schemas';
export const TEMP_SCHEMA_FILE_VERSION = 1;
//     /operations
const TEMP_OP_RECORDINGS_FOLDER_NAME = 'operations';
export const TEMP_OP_RECORDING_FILE_VERSION = 1;

// e.g. /grogqli
export const getRecordingsRootDir = async (): Promise<string> => {
  const config = await getConfig();
  const recordingsRootDir = config('recordingsSaveDirectory');
  await createDirIfDoesntExist(recordingsRootDir);
  return recordingsRootDir;
};

// e.g. /grogqli/schemas
export const getSchemasFolderPath = async (): Promise<string> => {
  const recordingsRootDir = await getRecordingsRootDir();
  const schemasFolderPath = path.join(recordingsRootDir, SCHEMAS_FOLDER_NAME);
  await createDirIfDoesntExist(schemasFolderPath);
  return schemasFolderPath;
};

// e.g. /grogqli/schemas/<schemaId>/operations.json
export const getOpNameMappingFilePath = async (schemaId: string) => {
  const schemasFolderPath = await getSchemasFolderPath();

  const opNameMappingFilePath = path.join(schemasFolderPath, schemaId);

  await createDirIfDoesntExist(opNameMappingFilePath);

  return path.join(
    opNameMappingFilePath,
    OPERATIONS_NAME_TO_ID_MAPPING_FILENAME
  );
};

// e.g. /grogqli/schemas/<schemaId>/types.json
export const getTypeNameMappingFilePath = async (schemaId: string) => {
  const schemasFolderPath = await getSchemasFolderPath();

  const typeNameMappingFilePath = path.join(schemasFolderPath, schemaId);

  await createDirIfDoesntExist(typeNameMappingFilePath);

  return path.join(typeNameMappingFilePath, TYPES_NAME_TO_ID_MAPPING_FILENAME);
};

// e.g. /grogqli/workflows
export const getWorkflowsFolderPath = async (): Promise<string> => {
  const recordingsRootDir = await getRecordingsRootDir();
  const workflowsFolderPath = path.join(
    recordingsRootDir,
    WORKFLOWS_FOLDER_NAME
  );
  await createDirIfDoesntExist(workflowsFolderPath);
  return workflowsFolderPath;
};

// e.g. /grogqli/workflows/index.json
export const getWorkflowNameMappingFilePath = async (): Promise<string> => {
  const workflowsFolderPath = await getWorkflowsFolderPath();
  return path.join(workflowsFolderPath, WORKFLOWS_NAME_TO_ID_MAPPING_FILENAME);
};

// e.g. /grogqli/workflows/<workflowId>.json
export const getWorkflowRecordingFilePath = async (
  workflowId: string
): Promise<string> => {
  const workflowsFolderPath = await getWorkflowsFolderPath();
  return path.join(workflowsFolderPath, `${workflowId}.json`);
};

// e.g. /grogqli/operations/<opId>.json
export const getOperationRecordingsFilePath = async (
  opId: string
): Promise<string> => {
  const recordingsRootDir = await getRecordingsRootDir();
  const pathToOperationRecordingFile = path.join(
    recordingsRootDir,
    OPERATIONS_FOLDER_NAME
  );

  await createDirIfDoesntExist(pathToOperationRecordingFile);

  return path.join(pathToOperationRecordingFile, `${opId}.json`);
};

// e.g. /grogqli/types/<typeId>.json
export const getTypeFilePath = async (typeId: string) => {
  const recordingsRootDir = await getRecordingsRootDir();

  const typeFilePath = path.join(recordingsRootDir, TYPES_FOLDER_NAME);

  await createDirIfDoesntExist(typeFilePath);

  return path.join(typeFilePath, `${typeId}.json`);
};

// e.g. /grogqli/local/schemas
export const getTemporarySchemaRecordingsFolderPath = async (): Promise<string> => {
  const recordingsRootDir = await getRecordingsRootDir();
  const schemaRecordingsPath = path.join(
    recordingsRootDir,
    TEMP_FOLDER_NAME,
    TEMP_SCHEMAS_FOLDER_NAME
  );
  await createDirIfDoesntExist(schemaRecordingsPath);
  return schemaRecordingsPath;
};

// e.g. /grogqli/local/schemas/<schemaHash>.json
type GetTemporarySchemaRecordingFilename = (
  schemaHash: string
) => Promise<string>;
export const getTemporarySchemaRecordingFilename: GetTemporarySchemaRecordingFilename = async (
  schemaHash
) => {
  const tempSchemaRecordingsPath = await getTemporarySchemaRecordingsFolderPath();
  return path.join(tempSchemaRecordingsPath, `${schemaHash}.json`);
};

// e.g. /grogqli/local/sessions/<sessionId>
const getSessionFolderPath = async (sessionId: string) => {
  const recordingsRootDir = await getRecordingsRootDir();

  const givenSessionsPersistencePath = path.join(
    recordingsRootDir,
    TEMP_FOLDER_NAME,
    HANDLER_SESSIONS_STATE_FOLDER_NAME,
    sessionId
  );

  await createDirIfDoesntExist(givenSessionsPersistencePath);

  return givenSessionsPersistencePath;
};

// e.g. /grogqli/local/sessions/<sessionId>/index.json
export const getSessionsFilePath = async (sessionId: string) => {
  const givenSessionsPersistencePath = await getSessionFolderPath(sessionId);

  return path.join(
    givenSessionsPersistencePath,
    HANDLER_SESSION_STATE_FILENAME
  );
};

// e.g. /grogqli/local/sessions/<sessionId>/operations
export const getTempOpRecordingsFolderPath = async (
  sessionId: string
): Promise<string> => {
  const givenSessionsPersistencePath = await getSessionFolderPath(sessionId);

  const tempOpRecordingFolderPath = path.join(
    givenSessionsPersistencePath,
    TEMP_OP_RECORDINGS_FOLDER_NAME
  );

  await createDirIfDoesntExist(tempOpRecordingFolderPath);

  return tempOpRecordingFolderPath;
};

// e.g. /grogqli/local/sessions/<sessionId>/operations/<tempOpRecordingId>.json
type GetTempOpRecordingFilePath = (params: {
  sessionId: string;
  tempOpRecordingId: string;
}) => Promise<string>;
export const getTempOpRecordingFileName: GetTempOpRecordingFilePath = async ({
  sessionId,
  tempOpRecordingId,
}) => {
  const tempOpRecordingFolderPath = await getTempOpRecordingsFolderPath(
    sessionId
  );

  return path.join(tempOpRecordingFolderPath, `${tempOpRecordingId}.json`);
};

// Schema
// ***
// **
// *
export interface SchemaFile {
  id: string;
  version: number;
  introspectionQuery: IntrospectionQuery;
}

export const getSchema = async (schemaId: string): Promise<SchemaFile> => {
  const schemasFolderPath = await getSchemasFolderPath();
  const absPathToSchema = path.join(schemasFolderPath, schemaId);

  await createDirIfDoesntExist(absPathToSchema);

  const absPathToSchemaFile = path.join(absPathToSchema, SCHEMA_FILENAME);

  return JSON.parse(await fs.promises.readFile(absPathToSchemaFile, 'utf8'));
};
