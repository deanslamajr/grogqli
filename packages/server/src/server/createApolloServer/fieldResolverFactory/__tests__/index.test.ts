import path from 'path';

import { fieldResolverFactory, ResolveValueFactoryParams } from '..';
import { Context } from '../..';
import { getConfig } from '../../../files/getConfig';
import {
  OperationRecordingsFile,
  OperationNameToIdMapping,
} from '../../../files/operation';
import { TypeNameToIdMappingVersion1 } from '../../../files/type';

import schema from './grogqli/schemas/o8eioU83f/schema.json';
import operationsData from './grogqli/schemas/o8eioU83f/operations.json';
import typeNameToIdMappingData from './grogqli/schemas/o8eioU83f/types.json';
import someOperationFile from './grogqli/operations/tRYadUuzxr.json';

jest.mock('../../../files/getConfig');
jest.mock('./grogqli/operations/tRYadUuzxr.json');

const workflowId = 'u33U7TWkJS';
const schemaId = 'o8eioU83f';
const opName = 'GetValues';
const opId = 'tRYadUuzxr';

describe('fieldResolverFactory', () => {
  let fieldResolverFactoryArgs: ResolveValueFactoryParams;
  let info: any;
  let context: Context;
  let mockedOperationFile: jest.Mocked<OperationRecordingsFile>;

  beforeEach(() => {
    const mockedGetConfig = getConfig as jest.MockedFunction<typeof getConfig>;
    mockedGetConfig.mockImplementation(async () => () => {
      const relativePathToTestGrogqli = path.normalize('./grogqli');
      return path.join(__dirname, relativePathToTestGrogqli);
    });

    mockedOperationFile = (someOperationFile as OperationRecordingsFile) as jest.Mocked<OperationRecordingsFile>;
    const actualOperationFile = jest.requireActual(
      `./grogqli/operations/${opId}.json`
    );
    mockedOperationFile.recordings = actualOperationFile.recordings;

    fieldResolverFactoryArgs = {
      schema: schema.introspectionQuery as any,
      operationsData: operationsData as OperationNameToIdMapping,
      parentTypeName: schema.introspectionQuery.__schema.queryType.name,
      fieldName: 'inputThings',
      typeNameToIdMappingData: typeNameToIdMappingData as TypeNameToIdMappingVersion1,
    };

    info = {
      operation: {
        name: {
          value: opName,
        },
      },
    };

    context = {
      schemaId,
      runTimeVariables: {
        grogqli: { workflowId },
      },
    };
  });

  it('should return a resolver function', () => {
    const resolveField = fieldResolverFactory(fieldResolverFactoryArgs);

    expect(typeof resolveField).toBe('function');
  });

  describe('a resolver function created from a fieldResolverFactory invocation', () => {
    describe('if an unnamed operation is resolved', () => {
      it('should throw', async () => {
        const infoWithUnnamedOp = {
          ...info,
          operation: {},
        };

        const resolveField = fieldResolverFactory(fieldResolverFactoryArgs);

        await expect(
          resolveField({}, {}, context, infoWithUnnamedOp)
        ).rejects.toThrowErrorMatchingSnapshot();
      });
    });

    describe('root type case', () => {
      it('should return the value associated with the given fieldName', async () => {
        const resolveField = fieldResolverFactory(fieldResolverFactoryArgs);

        const actual = await resolveField({}, {}, context, info);

        expect(actual).toMatchSnapshot();
      });

      describe('variables', () => {
        describe('resolving a field that does not have any associated args', () => {
          it('should correctly resolve the recording', async () => {
            throw new Error('Implement me!');
          });
        });

        describe('resolving a field that has associated arg(s)', () => {
          describe('for playback using MATCH_OR_ERROR strategy', () => {
            describe('if the given variables match a variable recording', () => {
              it.only('should resolve the recording associated with the matched variables recording', async () => {
                const args = {
                  input: {
                    type: 'MAC',
                  },
                };

                const resolveField = fieldResolverFactory(
                  fieldResolverFactoryArgs
                );

                const actual = await resolveField({}, args, context, info);

                expect(actual).toMatchSnapshot();
              });
            });

            describe('if the given variables DO NOT match a variable recording', () => {
              it('should throw an error', () => {
                throw new Error('Implement me!');
              });
            });

            describe('no variable recordings exist', () => {
              it('should throw an error', () => {
                throw new Error('Implement me!');
              });
            });
          });

          describe('for playback using MATCH_OR_DEFAULT strategy', () => {
            describe('if the given variables match a variable recording', () => {
              it('should resolve the recording associated with the matched variables recording', () => {
                throw new Error('Implement me!');
              });
            });

            describe('if the given variables DO NOT match a variable recording', () => {
              it('should resolve the default recording', () => {
                throw new Error('Implement me!');
              });
            });

            describe('no variable recordings exist', () => {
              it('should resolve the default recording', () => {
                throw new Error('Implement me!');
              });
            });
          });

          describe('for playback using ALWAYS_DEFAULT strategy', () => {
            describe('if the given variables match a variable recording', () => {
              it('should resolve the recording associated with the matched variables recording', () => {
                throw new Error('Implement me!');
              });
            });

            describe('if the given variables DO NOT match a variable recording', () => {
              it('should resolve the default recording', () => {
                throw new Error('Implement me!');
              });
            });

            describe('no variable recordings exist', () => {
              it('should resolve the default recording', () => {
                throw new Error('Implement me!');
              });
            });
          });
        });
      });

      describe('if a workflow file cannot be found for the given workflowId', () => {
        it('should throw', async () => {
          const resolveField = fieldResolverFactory(fieldResolverFactoryArgs);

          const contextWithNonexistentWorkflow = {
            ...context,
            runTimeVariables: {
              ...context.runTimeVariables,
              grogqli: {
                ...context.runTimeVariables.grogqli,
                workflowId: 'nonexistentWorkflowId',
              },
            },
          };

          await expect(
            resolveField({}, {}, contextWithNonexistentWorkflow, info)
          ).rejects.toThrowErrorMatchingSnapshot();
        });
      });

      describe('if the operations file does not have an recording for the given operation', () => {
        it('should throw', async () => {
          const infoWithUnrecordedOp = {
            ...info,
            operation: {
              name: {
                value: 'unrecordedOpName',
              },
            },
          };

          const resolveField = fieldResolverFactory(fieldResolverFactoryArgs);

          await expect(
            resolveField({}, {}, context, infoWithUnrecordedOp)
          ).rejects.toThrowErrorMatchingSnapshot();
        });
      });

      describe('if the workflow file does not have an entry for the given operation id', () => {
        it('should throw', async () => {
          const argsWithNonExistentOpId = {
            ...fieldResolverFactoryArgs,
            operationsData: {
              ...operationsData,
              operations: {
                ...operationsData.operations,
                someOpName: {
                  ...operationsData.operations.someOpName,
                  id: 'nonExistentOpId',
                },
              },
            } as OperationNameToIdMapping,
          };

          const resolveField = fieldResolverFactory(argsWithNonExistentOpId);

          await expect(
            resolveField({}, {}, context, info)
          ).rejects.toThrowErrorMatchingSnapshot();
        });
      });

      describe('if a type file cannot be found for the given root type', () => {
        it('should throw', async () => {
          const nonExistentRootTypeName = 'NonExistentRootType';
          const argsWithNonExistentRootType = {
            ...fieldResolverFactoryArgs,
            parentTypeName: nonExistentRootTypeName,
          };
          const resolveField = fieldResolverFactory(
            argsWithNonExistentRootType
          );

          await expect(
            resolveField({}, {}, context, info)
          ).rejects.toThrowErrorMatchingSnapshot();
        });
      });

      describe('if the root type file does not have a recording instance for the given recordingId', () => {
        it('should throw', async () => {
          mockedOperationFile.recordings = {
            ...mockedOperationFile.recordings,
            someOpRecordingId: {
              ...mockedOperationFile.recordings.someOpRecordingId,
              rootTypeRecordings: {
                ...mockedOperationFile.recordings.someOpRecordingId
                  .rootTypeRecordings,
                xhg87iT1: {
                  ...mockedOperationFile.recordings.someOpRecordingId
                    .rootTypeRecordings.xhg87iT1,
                  recordingId: 'nonExistentRecordingId',
                },
              },
            },
          };

          const resolveField = fieldResolverFactory(fieldResolverFactoryArgs);

          await expect(
            resolveField({}, {}, context, info)
          ).rejects.toThrowErrorMatchingSnapshot();
        });
      });

      describe('if the root type recording does not have an entry for the given fieldName', () => {
        it('should throw', async () => {
          const argsWithNonExistentFieldName = {
            ...fieldResolverFactoryArgs,
            fieldName: 'nonExistentFieldName',
          };
          const resolveField = fieldResolverFactory(
            argsWithNonExistentFieldName
          );

          await expect(
            resolveField({}, {}, context, info)
          ).rejects.toThrowErrorMatchingSnapshot();
        });
      });
    });

    describe('nested type case', () => {
      describe('variables', () => {
        describe('resolving a field that does not have any associated args', () => {
          it('should correctly resolve the recording', () => {
            throw new Error('Implement me!');
          });
        });

        describe('resolving a field that has associated arg(s)', () => {
          describe('for playback using MATCH_OR_ERROR strategy', () => {
            describe('if the given variables match a variable recording', () => {
              it('should resolve the recording associated with the matched variables recording', () => {
                throw new Error('Implement me!');
              });
            });

            describe('if the given variables DO NOT match a variable recording', () => {
              it('should throw an error', () => {
                throw new Error('Implement me!');
              });
            });

            describe('no variable recordings exist', () => {
              it('should throw an error', () => {
                throw new Error('Implement me!');
              });
            });
          });

          describe('for playback using MATCH_OR_DEFAULT strategy', () => {
            describe('if the given variables match a variable recording', () => {
              it('should resolve the recording associated with the matched variables recording', () => {
                throw new Error('Implement me!');
              });
            });

            describe('if the given variables DO NOT match a variable recording', () => {
              it('should resolve the default recording', () => {
                throw new Error('Implement me!');
              });
            });

            describe('no variable recordings exist', () => {
              it('should resolve the default recording', () => {
                throw new Error('Implement me!');
              });
            });
          });

          describe('for playback using ALWAYS_DEFAULT strategy', () => {
            describe('if the given variables match a variable recording', () => {
              it('should resolve the recording associated with the matched variables recording', () => {
                throw new Error('Implement me!');
              });
            });

            describe('if the given variables DO NOT match a variable recording', () => {
              it('should resolve the default recording', () => {
                throw new Error('Implement me!');
              });
            });

            describe('no variable recordings exist', () => {
              it('should resolve the default recording', () => {
                throw new Error('Implement me!');
              });
            });
          });
        });
      });
    });
  });
});
