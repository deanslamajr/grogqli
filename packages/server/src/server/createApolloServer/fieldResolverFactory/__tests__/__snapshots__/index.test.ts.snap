// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fieldResolverFactory a resolver function created from a fieldResolverFactory invocation if an unnamed operation is resolved should throw 1`] = `"TODO handle unnamed query case"`;

exports[`fieldResolverFactory a resolver function created from a fieldResolverFactory invocation root type case if a type file cannot be found for the given root type should throw 1`] = `
"
      TODO handle the case where the given typeName does not exist in the given types.json.
      typeName:NonExistentRootType
    "
`;

exports[`fieldResolverFactory a resolver function created from a fieldResolverFactory invocation root type case if a workflow file cannot be found for the given workflowId should throw 1`] = `"TODO handle case where a workflow file is not found for the given workflowId. workflowId:nonexistentWorkflowId"`;

exports[`fieldResolverFactory a resolver function created from a fieldResolverFactory invocation root type case if the operations file does not have an recording for the given operation should throw 1`] = `"TODO handle case where an opId does not exist for the given opName. opName:unrecordedOpName"`;

exports[`fieldResolverFactory a resolver function created from a fieldResolverFactory invocation root type case if the root type file does not have a recording instance for the given recordingId should throw 1`] = `"TODO Handle case where type file does not have a recording for the given id. typeId:xhg87iT1 recordingId:nonExistentRecordingId"`;

exports[`fieldResolverFactory a resolver function created from a fieldResolverFactory invocation root type case if the root type recording does not have an entry for the given fieldName should throw 1`] = `
"TODO handle case where recording doesn't have a value for the given field.
          parentTypeName:Query fieldName:nonExistentFieldName"
`;

exports[`fieldResolverFactory a resolver function created from a fieldResolverFactory invocation root type case if the workflow file does not have an entry for the given operation id should throw 1`] = `"TODO handle case where the workflow file does not have a recording for the given opId. opId:nonExistentOpId"`;

exports[`fieldResolverFactory a resolver function created from a fieldResolverFactory invocation root type case should return the value associated with the given fieldName 1`] = `"someAccountRecordingId"`;

exports[`fieldResolverFactory a resolver function created from a fieldResolverFactory invocation root type case variables resolving a field that has associated arg(s) for playback using MATCH_OR_ERROR strategy if the given variables match a variable recording should resolve the recording associated with the matched variables recording 1`] = `
Array [
  "92gGfleoXJ",
  "ccJBAicioA",
  "twkCcrImXy",
  "nYsDKRgx94",
  "txPmta9CXW",
  "EJEPkJdh4g",
  "NhDTNsmtYl",
  "hUXCIV8PmO",
  "esFHu5FLdz",
  "IV6EFY7HTf",
]
`;
